source/common/grpc/common.cc

```diff
@@ -269,5 +286,98 @@ std::string Common::typeUrl(const std::string& qualified_name) {
   return typeUrlPrefix() + "/" + qualified_name;
 }
 
+struct BufferInstanceContainer {
+  BufferInstanceContainer(int ref_count, Buffer::InstancePtr buffer)
+      : ref_count_(ref_count), buffer_(std::move(buffer)) {}
+  std::atomic<int> ref_count_;
+  Buffer::InstancePtr buffer_;
+};
+
+static void derefBufferInstanceContainer(void* container_ptr) {
+  auto container = reinterpret_cast<BufferInstanceContainer*>(container_ptr);
+  container->ref_count_--;
+  if (container->ref_count_ <= 0) {
```

Comment:
Maybe explain this is thread safe as we never increment the ref count in a comment. It might be easier to figure out a clever way to do this with shared_ptr/weak_ptr or something similar.

Reply:
The problems is that the API provided by grpc::Slice only provides for a raw pointer and a function pointer (not a std::function).  I am open to suggestions if you have a way to do that.